tokio/src/sync/mpsc/chan.rs

```diff
@@ -241,6 +241,24 @@ impl<T, S: Semaphore> Rx<T, S> {
         self.inner.notify_rx_closed.notify_waiters();
     }
 
+    pub(crate) fn is_closed_and_empty(&mut self) -> bool {
+        // There two internal states that can represent a closed channel that has no messages
+        //
+        //  1. When `close` is called.
+        //  In this case, the inner semaphore will be closed, and the channel has no
+        //  outstanding messages if the list is at the tail position.
+        //
+        //  2. When all senders are dropped.
+        //  In this case, the semaphore remains unclosed, and the `index` in the list won't
+        //  reach the tail position. It is necessary to check the list if the last block is
+        //  `closed` and has no value available.
+        self.inner.rx_fields.with_mut(|rx_fields_ptr| {
+            let rx_fields = unsafe { &mut *rx_fields_ptr };
+            (self.inner.semaphore.is_closed() && rx_fields.list.is_at_tail_position(&self.inner.tx))
+                || rx_fields.list.is_closed_and_empty(&self.inner.tx)
```

Comment:
On alternative to this implementation would be to change the `Tx` drop implementation to also close the semaphore, this would simplify the internal state to rely only on the semaphore.

The issue with this alternative approach is that is necessary to add a constraint to the struct generic parameter.

```rust
impl<T, S: Semaphore> Drop for Tx<T, S> { // <- Tx<T, S: Semaphore>, this would propagate to a lot of places
    fn drop(&mut self) {
        if self.inner.tx_count.fetch_sub(1, AcqRel) != 1 {
            return;
        }

        // Close the list, which sends a `Close` message
        self.inner.tx.close();
        self.inner.semaphore.close();

        // Notify the receiver
        self.wake_rx();
    }
}
```